ESPS/general/src/srecord/srecord.1
.\" Copyright (c) 1992 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)srecord.1	1.5 3/28/97 ERL
.ds ]W (c) 1992 Entropic Research Laboratory, Inc.
.TH SRECORD 1\-ESPS 10/24/92
.SH NAME
srecord - record to disk or pipe for SUN SPARC built-in audio device
.PP
.SH SYNOPSIS
.B srecord
[
.BI \-s " duration"
] [
.BI \-S
] [
.BI \-W " xwaves display args."
] [
.BI \-P
] [
.BI \-p " prompt string"
] [
.BI \-x " debug-level"
] [
.BI \-H
] [
.BI \-e
] [[
.BI \-o
]
.I file
]
.PP
.SH DESCRIPTION
.PP
.PP
\fISrecord\fP provides recording to local disk using SUN SPARCstation
built-in audio device, sampling at 8000 Hz.
Direct recording onto network disks is often
feasible as well.  Output data files are linear encoded and have ESPS FEA_SD 
headers, or,
optionally, no headers.  Output may optionally be directed to
\fIstdout\fP.  \fISrecord\fP has special adaptations that permit tight
coupling with \fIxwaves\fP (see \fIINTERACTION WITH XWAVES\fP below).
.PP
Note that processes consuming the output of
\fIsrecord\fP on a pipe must be able to keep up with the average aggregate
sampling frequency.  Options are available to control the recording duration, prompting, header suppression, and immediate display by \fIxwaves\fP.
.PP
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-s " duration" "\fR [10]"
Specifies the maximum duration of the recording session in seconds.
Recording may be interrupted before this time expires with SIGINT,
SIGQUIT or SIGUSR1 (see below).  The default \fIduration\fP is 10 seconds.
The upper limit is set only by disk space.
.TP
.BI \-S
Enable the \fIxwaves\fP(1\-ESPS) "make" command via \fIsend_xwaves2\fP(3\-ESPS) when the
requested recording time has elapsed or when recording is interrupted.  This 
permits immediate examination of the recorded passage using \fIxwaves\fP.  
See \fIINTERACTION WITH XWAVES\fP below.
.TP
.BI \-W
The argument to this option will be appended to the \fIsend_xwaves\fP
"make" command to permit display customization (e.g. via
window location and size specifications).  See \fIINTERACTION
WITH XWAVES\fP below.
.TP
.BI \-P
Enable a prompt message when A/D has actually commenced.  The default message
is a "bell ring" and the text "START RECORDING NOW...."  This prompt may be changed
with the \-p option.
.TP
.BI \-p " prompt string"
\fIPrompt string\fP will be used as the alert that recording is commencing.  Specifying
\-p forces -P.
.TP
.BI \-H
Suppresses header creation.  A "bare" sample stream will result.  The default is
to produce an ESPS FEA_SD file.
.TP
.BI \-e
This option causes a/d input to be echoed out the d/a during recording.
.TP
.BI \-x " debug_level"
Setting debug_level nonzero causes several messages to be printed as
internal processing proceeds.  The default is level 0, which causes no debug
output.
.TP
.BI -o " output"
Specifies a file for output.  Use of -\fIo\fP before the file designation is optional
if the filename is not the last command-line component.
.PP
.SH "INTERACTION WITH XWAVES"
.PP
\fISrecord\fP is designed to optionally use the server mode of \fIxwaves\fP(1\-ESPS) for
display of its output file on completion of the record operation.
This is implemented using \fIsend_xwaves2\fP(3\-ESPSu).  The following
conditions must be met for this feature to work.  (1) \fIXwaves\fP must be
running in the server mode. (2) The environment variables WAVES_PORT
and WAVES_HOST must be correctly defined (see \fIespsenv\fP(1\-ESPS)).  (3)
The record operation must be interrupted with a SIGUSR1 signal (e.g.
via "kill \-30 pid," where pid is the process ID of the \fIsrecord\fP
process), or if \fIsrecord\fP is not thus interrupted, the -\fIS\fP flag must have
been set.  (4) Output must be to a file.
.PP
An example \fImbuttons\fP(1\-ESPS) script to implement a primitive record
control panel follows:
.PP
.nf
.na
.ne 10
"RECORD"	exec srecord  -P -s60 -S -W"name $$ loc_y 150" \\
                xx$$& echo $! > foo
"STOP"          kill -30 `cat foo`
"ERASE"         f=`cat foo` ; k=`echo $f 1 - p q | dc` ; \\
		kill -2 $f ; rm -f xx$k ; send_xwaves kill name $k
.fi
.ad
.PP
Note how the -\fIW\fP option is used to name the display ensemble and to fix
the vertical location of the waveform at the same place on consecutive
invocations.  In general, the -\fIW\fP option can be used to augment the
display generation as described under the "make" command in the \fIxwaves\fP
manual.  Note that the "STOP" function is implemented with a "kill
-30" (SIGUSR1).  This causes \fIsrecord\fP to send the "make" command to
\fIxwaves\fP.  If either kill -2 (SIGINT) or kill -3 (SIGQUIT) is sent to
\fIsrecord\fP, it will terminate gracefully, but will not send any messages
to \fIxwaves\fP.  The -\fIS\fP option causes the \fIxwaves\fP display operation to occur
even in the non-interrupted case (i.e. after 60 sec of recording).
The above script is not robust, but may serve as a useful starting
point for more serious attempts.
.PP
.SH ESPS PARAMETERS
.PP
The parameter file is not read.  
.PP
.SH ESPS COMMON
.PP
ESPS Common is not read or written.
.PP
.SH WARNINGS
.PP
When output is to a file, the ESPS header, if it is present, will
correctly reflect the absolute maximum sample value encountered during
recording and the number of samples recorded.  If output is to a pipe,
these values are not recorded in the header.
.PP
If another /fIsrecord/fR process is started while one is in progress,
it will hang until the device becomes available.  Use the /fI-P/fR option
to signal when the recording starts.
.PP
.SH SEE ALSO
.PP
.nf
FEA_SD(5\-\s-1ESPS\s+1), \fItestsd\fP(1\-\s-1ESPS\s+1), \fIcopysd\fP(1\-\s-1ESPS\s+1),
\fIsplay\fP(1\-\s-1ESPS\s+1), \fIsfconvert\fP(1\-\s-1ESPS\s+1), \fIsetmax\fP(1\-\s-1ESPS\s+1),
\fImu2esps\fP(1\-\s-1ESPS\s+1)
.fi
.PP
.SH AUTHOR
.PP
Derek Lin


ESPS/general/src/srecord/srecord.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1990-1992 Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Derek Lin
 * Checked by:
 * Revised by:
 *
 * Brief description:
 *
 */

static char *sccs_id = "@(#)srecord.c	1.7	6/23/93	ERL";


#include <stdio.h>
#include <math.h>
#include <fcntl.h>
#include <signal.h>
#include <esps/esps.h>
#include <esps/epaths.h>
#include <esps/ss.h>

#define SPARC_SAMPLE_FREQ 8000
#define SIZ_CHANNEL 1
#define SYNTAX { \
fprintf(stderr,"Usage: %s [-x<debug level>] [-H] [-P] [-p<prompt string>] [-S] [-W<for waves make>] [-s<duration>] [-e] [[-o] output file]\n",av[0]); \
  fprintf(stderr, "Duration is in seconds and defaults to %f.\n",duration); \
  fprintf(stderr, "-H suppresses header creation, else an ESPS header is prepended.\n"); \
  fprintf(stderr, "-P enables prompting at start of A/D conversion\n"); \
  fprintf(stderr, "-p specifies the prompt to be used at start of A/D conversion\n"); \
  fprintf(stderr, "-S enables sending of the waves load file on completion.\n"); \
  fprintf(stderr, "The argument of -W gets appended to the send_xwaves make command.\n");  \
  fprintf(stderr, "-e enables echo of a/d input to d/a.\n"); \
}		  

/*
  This program records using SUN codec, and writes the ouput to stdout or file.
  The output file can optionally have no header or an ESPS header.
*/


#define ERR 4

/* BUF_SIZE is the maximum size of each UNIX I/O operation.  Below,
   this buffer size is tied to the physical space available on the S-bus
   board.  */

#define BUF_SIZE NELEMS

int debug_level = 0, send_display_file = FALSE, do_prompt = FALSE;
char *pstring = "START RECORDING NOW ...";

/*---------------------------------------------------------------------*/
output_the_header( stream, freq, amax, samples, channel)
     FILE *stream;
     double amax, freq;
     int samples, channel;
{
  double start_time = 0.0;
  struct header *head;
  void set_pvd();

  if(stream && (freq > 0.0) && (samples > 0)) {
    head = new_header(FT_FEA);
    set_pvd(head);
    (void)init_feasd_hd(head, SHORT, (channel < 2)?1:2, &start_time, NO, freq);
    if(stream != stdout) {
      *add_genhd_d("max_value", (double *)NULL, 1, head) = amax;
      *add_genhd_f("samples", (float *)NULL, 1, head) = samples;
    }
    (void)write_header(head, stream);
  } else
    fprintf(stderr,"Bad args to output_the_header(stream=%x freq=%f samples=%d\n",
	      stream, freq, samples);
  return(NULL);
}


/*---------------------------------------------------------------------*/
main(ac, av)
     int ac;
     char **av;
{
  int  channel = SIZ_CHANNEL;
  int echo_to_da = 0;
  double duration = 10.0, freq = SPARC_SAMPLE_FREQ;
  char *ofile = NULL;
  char *display_stuff = NULL;
  FILE *output = stdout;
  int c, do_header = TRUE;
  extern int optind;
  extern char *optarg;
  int i;

  while((c = getopt(ac,av,"x:o:s:W:p:SPH:?e")) != EOF) {
    switch(c) {
    case 'S':
      send_display_file = TRUE;
      break;
    case 'W':
      display_stuff = optarg;
      break;
    case 'P':
      do_prompt = TRUE;
      break;
    case 'p':
      pstring = optarg;
      do_prompt = TRUE;
      break;
    case 's':
      duration = atof(optarg);
      break;
    case 'x':
      debug_level = atoi(optarg);
      break;
    case 'o':
      ofile = optarg;
      if(!(output = fopen(ofile,"w"))) {
	fprintf(stderr,"Can't open %s for output.\n",ofile);
	exit(ERR);
      }
      break;
    case 'H':
      do_header = FALSE;
      break;
    case 'e':
      echo_to_da = 1;
      break;
    case '?':
    default:
      SYNTAX
    }
  }
  if(ac < 2){
    SYNTAX
    exit(1);
  }

  if(ac > optind) {
    ofile = av[optind];
    if(strcmp(ofile,"-")) {
      if(!(output = fopen(ofile,"w"))) {
	fprintf(stderr,"Can't open %s for output.\n",ofile);
	exit(ERR);
      }
    } else {
      optind++;
      ofile = NULL;
      output = stdout;
    }
  }
    
 {
   int samples_requested = 0.5 + (duration*freq), got;
   double amax = 0.0;

   if(do_header)
       output_the_header(output, freq, amax, samples_requested, channel);
   
   got = record_file(output, samples_requested, &freq, channel, 
	&amax, echo_to_da);
   if(got) {
     if(do_header && (output != stdout)) { /* rewind and update header values*/
       fseek(output, 0, 0);
       output_the_header(output, freq, amax, got);
       fseek(output, 0, 2);
     }
     fclose(output);
     if(ofile && (output != stdout) && send_display_file) {
       char mess[500];
       if(display_stuff != NULL)
	 sprintf(mess,"make file %s %s\n", ofile, display_stuff);
       else
	 sprintf(mess,"make file %s\n", ofile);
       send_xwaves2(NULL, NULL, mess, debug_level);
     }
     exit(0);
   } else
     fprintf(stderr,"No samples collected!\n");
   exit(ERR);
 }
}


char	*ProgName = "srecord";
char	*Version = "1.3";
char	*Date = "6/10/93";

void
set_pvd(hdr)
    struct header   *hdr;
{
    (void) strcpy(hdr->common.prog, ProgName);
    (void) strcpy(hdr->common.vers, Version);
    (void) strcpy(hdr->common.progdate, Date);
}
